homeassistant/components/landisgyr_heat_meter/sensor.py

```diff
@@ -56,38 +76,52 @@ def __init__(self, coordinator, description, device):
         self.entity_description = description
 
         self._attr_device_info = device
-        self._attr_should_poll = bool(self.key in ("heat_usage", "heat_previous_year"))
-
-    async def async_added_to_hass(self) -> None:
-        """Call when entity about to be added to hass."""
-        await super().async_added_to_hass()
-        state = await self.async_get_last_sensor_data()
-        if state:
-            self._attr_native_value = state.native_value
+        self.update_values_from_api()
+
+    def update_values_from_api(self):
+        """Update attribute values with the data from the api coordinator."""
+        if self.coordinator.data is None:
+            _LOGGER.debug(
+                "Cannot update %s: could not read any data from the Heat Meter",
+                self.key,
+            )
+            return
 
-    def _handle_coordinator_update(self) -> None:
-        """Handle updated data from the coordinator."""
-        if self.key in asdict(self.coordinator.data):
+        api_key = API_KEY[self.key]
+        if api_key in asdict(self.coordinator.data):
             if self.device_class == SensorDeviceClass.TIMESTAMP:
                 self._attr_native_value = dt_util.as_utc(
-                    asdict(self.coordinator.data)[self.key]
+                    asdict(self.coordinator.data)[api_key]
                 )
             else:
-                self._attr_native_value = asdict(self.coordinator.data)[self.key]
-
-        if self.key == "heat_usage":
-            self._attr_native_value = convert_gj_to_mwh(
-                self.coordinator.data.heat_usage_gj
-            )
+                self._attr_native_value = asdict(self.coordinator.data)[api_key]
```

Comment:
This is better to be moved into the entity description (e.g., by adding `state_fn` property into those).

Reply:
Why is `asdict` needed here?

Reply:
> This is better to be moved into the entity description (e.g., by adding `state_fn` property into those).
I'll look into this.

> Why is `asdict` needed here?
Perhaps it isn't. What would be a better way?

Reply:
> What would be a better way?

I don't know. I'm not aware what the source type/data is, hence me asking why it needs `asdict`?

Reply:
> I don't know. I'm not aware what the source type/data is, hence me asking why it needs `asdict`?

The response is a HeatMeterResponse class which is a dataclass, allowing self.coordinator.data.heat_usage_gj (for example), but with the types dynamically stored in HEAT_METER_SENSOR_TYPES, this is not possible. The asdict solves this.

Reply:
I've rewritten this to use getattr, which is much nicer I think.